### PR TITLE
script to pre-validate and report on vDU nodes

### PR DESCRIPTION
**Pull request by dcritch**
Currently works well against HPE systems, based on a recent requirement working in a partner's lab. All the BIOS values they shared for HPE are tested, but mapping all the OpenShift recommended ones is an ongoing process.

Will add other systems as I get access and can test. There is very little consistencies between redfish implementations which is a bummer.

### CODE REVIEW

File: tooling/vdu-prechecks/vdu_precheck.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    }
+}
+
+supported_nics = ['E810', 'XXV710']
+
+dpp = pprint.PrettyPrinter(indent=2)
+
+def get_url(url):
+    '''
+    retrives a URL and returns the json formatted response
```
Comment (karmab) [inline tooling/vdu-prechecks/vdu_precheck.py]: retrieves

File: tooling/vdu-prechecks/vdu_precheck.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    '''
+
+    nics = []
+
+    if vendor == "HPE":
+        url = f"https://{args.ip}{sep}/BaseNetworkAdapters"
+    elif vendor == "Dell Inc.":
+        url = f"https://{args.ip}{sep}/NetworkAdapters"
+    else:
+        print(f"unsupported server vendor: {vendor}")
```
Comment (karmab) [inline tooling/vdu-prechecks/vdu_precheck.py]: Unsupported

File: tooling/vdu-prechecks/vdu_precheck.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+                for mac in port_data['AssociatedNetworkAddresses']:
+                    macs.append(mac)
+                nics.append({
+                    'model': nic_model,
+                    'slot': nic_slot,
+                    'macs': macs
+                })
+            return nics
+
+    return None
```
Comment (karmab) [inline tooling/vdu-prechecks/vdu_precheck.py]: you should return [] instead (you dont need to return None btw, it s the default behaviour)

File: tooling/vdu-prechecks/vdu_precheck.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+def supported_nic(model):
+    '''
+    test if a given NIC is supported
+    '''
+
+    for supported in supported_nics:
+        if supported in model:
+            return True
+
+    return False
```
Comment (karmab) [inline tooling/vdu-prechecks/vdu_precheck.py]: you can make it a oneliner with
return len([supported in model for supported in supported_nics]) > 0

File: tooling/vdu-prechecks/vdu_precheck.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+            slot = drive_data['Name']
+            disks.append({
+                'model': model,
+                'type': media_type,
+                'serial': serial,
+                'slot': slot
+            })
+
+    if len(disks) == 0:
+        print("no disks found")
```
Comment (karmab) [inline tooling/vdu-prechecks/vdu_precheck.py]: No disks

File: tooling/vdu-prechecks/vdu_precheck.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+def main():
+    '''
+    report on various aspects of a system to determine if its ready
+    to run a vDU workload
+    '''
+
+    sep = get_endpoint()
+    vendor = get_vendor(sep)
+    print(f"checking BIOS for {vendor} system @ {args.ip}")
```
Comment (karmab) [inline tooling/vdu-prechecks/vdu_precheck.py]: Checking

File: tooling/vdu-prechecks/vdu_precheck.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+
+    if vendor in reference_bios:
+        check_bios(sep, vendor)
+    else:
+        print(f"no reference BIOS config for {vendor}")
```
Comment (karmab) [inline tooling/vdu-prechecks/vdu_precheck.py]: No reference

File: tooling/vdu-prechecks/README.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+queries redfish to determine if a system is ready to run a vDU workload
+
+options:
+  -h, --help            show this help message and exit
+  -i IP, --ip IP        redfish ip
+  -u USER, --user USER  redfish user
+  -p PASSWORD, --password PASSWORD
+                        redfish password
+  -t TYPE, --type TYPE  vdu type (mb vs. lb)
```
Comment (karmab) [inline tooling/vdu-prechecks/README.md]: can you explain somewhere what mb and lb mean?

File: tooling/vdu-prechecks/vdu_precheck.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+supported_nics = ['E810', 'XXV710']
+
+dpp = pprint.PrettyPrinter(indent=2)
+
+def get_url(url):
+    '''
+    retrives a URL and returns the json formatted response
+    '''
+
+    request = requests.get(url, verify=args.verify, auth=(args.user, args.password))
```
Comment (karmab) [inline tooling/vdu-prechecks/vdu_precheck.py]: That seems to be the only line using requests.
If so, you can use the following to remove the need for an extra dependency

```
import base64
import ssl
from urllib.request import urlopen, Request

headers = {'Content-Type': 'application/json', 'Accept': 'application/json'}
credentials = base64.b64encode(bytes(f'{user}:{password}', 'ascii')).decode('utf-8')
headers["Authorization"] = f"Basic {credentials}
context = ssl.create_default_context()
context.check_hostname = False
context.verify_mode = ssl.CERT_NONE
request = Request(url, headers=headers)
return json.loads(urlopen(request, context=context).read())
```

**Review by karmab — APPROVED** [review]
LGTM, some stylistic snobbish feedback but make what you want with them :)

**Comment by dcritch** [comment]
Thanks for the snobby feedback! Good call on reducing dependencies, and I found a couple of other lowercase letters you must have missed :)

File: tooling/vdu-prechecks/vdu_precheck.py (hunk 3 above)
Comment (dcritch) [reply tooling/vdu-prechecks/vdu_precheck.py:194]: pylint barks at me if I remove it or return []:
~~~
R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
~~~

Comment (dcritch) [reply tooling/vdu-prechecks/vdu_precheck.py]: nevermind... I see what you mean. Fixed.

**Pull request merged**